docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway.
+There are a few specific scenarios that will be perf-sensitive, and an overload for `GetSemanticModel()` that skips nullability information will be provided for these.
+It will share an initial-binding `BoundNode` cache with the standard semantic model, so it will get the benefits if something else with the same semantic model requests information, as it would have today.
+
+## New APIs
+
+We add the new APIs for retreiving and interacting with nullability:
+
+```C#
+public enum Nullability
+{
+    NotApplicable,
+    NotComputed,
+    Unknown,
+    NotNull,
+    MayBeNull
+}
+
+#region Declared Nullability
+
+// These can potentially be influenced by calculated flow state, but do not change after initial declaration
+
+public interface IDiscardSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IEventSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IFieldSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface ILocalSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IMethodSymbol : ISymbol
+{
+    Nullability ReturnNullability { get; }
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+    Nullability ReceiverNullability { get; }
+}
+
+public interface IParameterSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+#endregion
+
+#region Flow State Nullability
+
+public interface IArrayTypeSymbol : ITypeSymbol
+{
+    Nullability ElementNullability { get; }
+}
+
+public interface INamedTypeSymbol : ITypeSymbol
+{
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+}
+
+public interface ITypeParameterSymbol : ITypeSymbol
+{
+    Nullability ReferenceTypeConstraintNullability { get; }
+    ImmutableArray<Nullability> ConstraintsNullabilities { get; }
+}
+
+public struct TypeInfo
+{
+    public Nullability Nullability { get; }
+    public Nullability ConvertedNullability { get; }
+}
+
+#endregion
+
+#region Compilation
+
+public class Compilation
+{
+    public SemanticModel GetSemanticModel(bool skipNullabilityInformation);
+}
+
+
+#endregion
+
+#region Printing
+
+public interface ITypeSymbol
+{
+    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
```

Comment:
this feels like the tip of the iceberg.  Because ITypeSymbol doesn't just know if it's nullable, this information must be passed along.  Note: this is viral.  Any method that needs to bottom out here will need to pass tha tinformation along.  Similarly, many existing methods will now need to take ITypeSymbols *and* this new 'Nullability' info.  This too will be viral, necessitating large swaths of code be updated accordingly.  For existing APIs that are ITypeSymbol-based, this seems like this will effectively be deprecating things.